Fixing DISTINCT with AS function

xiangfu0 opened the pull request:
## Description
To support queries like: 
```
select CAST(runs AS string) as ddd from baseballStats GROUP BY 1
select DISTINCT(CAST(runs AS string) as ddd) from baseballStats
```

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
<!-- If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release. -->

<!-- If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text.
-->
## Documentation
<!-- If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document
-->

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/TransformFunctionFactory.java:
```diff
@@ -198,6 +199,9 @@ public static TransformFunction get(ExpressionContext expression, Map<String, Da
         FunctionContext function = expression.getFunction();
         String functionName = canonicalize(function.getFunctionName());
         List<ExpressionContext> arguments = function.getArguments();
+        if (functionName.equalsIgnoreCase("AS")) {
+          arguments = arguments.subList(0, 1);
+        }
```
why not just pass both of them in? it shouldn't matter as long as `AsTransformFunction` handles it correctly yes?

xiangfu0 replied:
cause the other side is an identifier but not defined in schema, it will cause underly data block initialization error with NPE.

walterddr reviewed (approved):
looks good. plz fix checkstyle.

Jackie-Jiang commented:
I don't think we should introduce `AS` as a transform function. The fix should be purely on the query rewrite.

`select CAST(runs AS string) as ddd from baseballStats GROUP BY 1` should be rewritten to `select DISTINCT(CAST(runs AS string)) as ddd from baseballStats` instead of `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats`. Notice the place of `as ddd`.

In fact, `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats` is not a valid SQL query

xiangfu0 commented:
> I don't think we should introduce `AS` as a transform function. The fix should be purely on the query rewrite.
> 
> `select CAST(runs AS string) as ddd from baseballStats GROUP BY 1` should be rewritten to `select DISTINCT(CAST(runs AS string)) as ddd from baseballStats` instead of `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats`. Notice the place of `as ddd`.
> 
> In fact, `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats` is not a valid SQL query

True, I think the tricky case here is that distinct can have multiple arguments, e.g. 
```
select CAST(runs AS string) as a, CAST(num AS int) as b from baseballStats GROUP BY 1, 2
```

walterddr commented:
i think the fundamental problem here is that `distinct` is modeled as a function. where distinct is actually a selectNode attribute in calcite see: https://github.com/apache/calcite/blob/82dd78a14f6aef2eeec2f9c94978d04b4acc5359/core/src/main/java/org/apache/calcite/sql/SqlSelect.java#L140-L142.

For example as @Jackie-Jiang mentioned: 
```
select CAST(runs AS string) as a, CAST(num AS int) as b from baseballStats GROUP BY 1, 2
```
actually should be modified as:
```
SELECT DISTINCT -- notice that DISTINCT is a modifier of SELECT, not the 2 projection elements.
  CAST(runs AS string) as a, 
  CAST(num AS int) as b
FROM
  baseballStats 
GROUP BY 1, 2
```
see calcite syntactic note:
```
select:
      SELECT [ hintComment ] [ STREAM ] [ ALL | DISTINCT ]
          { * | projectItem [, projectItem ]* }
      FROM tableExpression
      [ WHERE booleanExpression ]
      [ GROUP BY { groupItem [, groupItem ]* } ]
      [ HAVING booleanExpression ]
      [ WINDOW windowName AS windowSpec [, windowName AS windowSpec ]* ]
```
DISTINCT is a modifier of the SELECT, not the underlying projectItems.

Jackie-Jiang commented:
> > I don't think we should introduce `AS` as a transform function. The fix should be purely on the query rewrite.
> > `select CAST(runs AS string) as ddd from baseballStats GROUP BY 1` should be rewritten to `select DISTINCT(CAST(runs AS string)) as ddd from baseballStats` instead of `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats`. Notice the place of `as ddd`.
> > In fact, `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats` is not a valid SQL query
> 
> True, I think the tricky case here is that distinct can have multiple arguments, e.g.
> 
> ```
> select CAST(runs AS string) as a, CAST(num AS int) as b from baseballStats GROUP BY 1, 2
> ```

For this query, can we rewrite it to `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats`? This way it is still valid SQL

xiangfu0 commented:
> > > I don't think we should introduce `AS` as a transform function. The fix should be purely on the query rewrite.
> > > `select CAST(runs AS string) as ddd from baseballStats GROUP BY 1` should be rewritten to `select DISTINCT(CAST(runs AS string)) as ddd from baseballStats` instead of `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats`. Notice the place of `as ddd`.
> > > In fact, `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats` is not a valid SQL query
> > 
> > 
> > True, I think the tricky case here is that distinct can have multiple arguments, e.g.
> > ```
> > select CAST(runs AS string) as a, CAST(num AS int) as b from baseballStats GROUP BY 1, 2
> > ```
> 
> For this query, can we rewrite it to `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats`? This way it is still valid SQL

I think this query `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats`  actually changes the number of output columns?

Jackie-Jiang commented:
> > > > I don't think we should introduce `AS` as a transform function. The fix should be purely on the query rewrite.
> > > > `select CAST(runs AS string) as ddd from baseballStats GROUP BY 1` should be rewritten to `select DISTINCT(CAST(runs AS string)) as ddd from baseballStats` instead of `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats`. Notice the place of `as ddd`.
> > > > In fact, `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats` is not a valid SQL query
> > > 
> > > 
> > > True, I think the tricky case here is that distinct can have multiple arguments, e.g.
> > > ```
> > > select CAST(runs AS string) as a, CAST(num AS int) as b from baseballStats GROUP BY 1, 2
> > > ```
> > 
> > 
> > For this query, can we rewrite it to `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats`? This way it is still valid SQL
> 
> I think this query `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats` actually changes the number of output columns?

This should have the same behavior as `SELECT C1 AS ALIAS_C1, C2 AS ALIAS_C2, ADD(alias_c1, alias_c2) FROM Foo` which you just fixed in #7590. I think the alias part will be ignored

xiangfu0 commented:
> > > > > I don't think we should introduce `AS` as a transform function. The fix should be purely on the query rewrite.
> > > > > `select CAST(runs AS string) as ddd from baseballStats GROUP BY 1` should be rewritten to `select DISTINCT(CAST(runs AS string)) as ddd from baseballStats` instead of `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats`. Notice the place of `as ddd`.
> > > > > In fact, `select DISTINCT(CAST(runs AS string) as ddd) from baseballStats` is not a valid SQL query
> > > > 
> > > > 
> > > > True, I think the tricky case here is that distinct can have multiple arguments, e.g.
> > > > ```
> > > > select CAST(runs AS string) as a, CAST(num AS int) as b from baseballStats GROUP BY 1, 2
> > > > ```
> > > 
> > > 
> > > For this query, can we rewrite it to `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats`? This way it is still valid SQL
> > 
> > 
> > I think this query `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats` actually changes the number of output columns?
> 
> This should have the same behavior as `SELECT C1 AS ALIAS_C1, C2 AS ALIAS_C2, ADD(alias_c1, alias_c2) FROM Foo` which you just fixed in #7590. I think the alias part will be ignored

oh, i mean `select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats`  will output 4 columns: a, b, a, b. but we only want 2 columns a, b.

walterddr commented:
hmm. I think we are off from the SQL semantics. 

1. first of all. DISTINCT is not a function, it is not the same as 
```
SELECT C1 AS ALIAS_C1, C2 AS ALIAS_C2, ADD(alias_c1, alias_c2)
```
because it doesn't produces a projection/transformation column for each input row.

2. this following SQL 
```
select CAST(runs AS string) as a, CAST(num AS int) as b, DISTINCT(a, b) from baseballStats
```
is only valid b/c DISTINCT is modeled as a function, and based on this precondition, the result cannot be `a,b,a,b` because you cannot have the first `a,b` modeled as non-distinct columns and the second `a,b` as distinct columns --> it simply isn't a valid linear algebra --> Imagine you have `SELECT a, DISTINCT(b) FROM table`.
which `a` value should I return if there's 2 rows with the same `b` value?

The fundamental problem here is that DISTINCT should not be modeled as a function.

amrishlal commented:
> 
> The fundamental problem here is that DISTINCT should not be modeled as a function.
>

True. DISTINCT is a keyword in SQL that should apply to the entire output row, but not to specific values within the row. The DISTINCT query should look like:

`SELECT DISTINCT a AS col1, b AS col2 FROM table.`

but not

`SELECT DISTINCT(a, b) AS col3 FROM table.`

because the latter can be abused, as you described, to write semantically incorrect statements such as:

`SELECT DISTINCT(a), b FROM table.`

Internally though, there is nothing wrong in rewriting query with DISTINCT keyword into a query with DISTINCT function provided that we make sure that the DISTINCT function includes all expressions from the select list and passes the alias associated with those expressions as well. Externally in SQL, the use of DISTINCT function should either be blocked or deprecated or at the very least restricted so that one cannot specify columns or expressions in the select list that are outside the DISTINCT function if one exists in the select list.

amrishlal commented on pinot-common/src/test/java/org/apache/pinot/sql/parsers/rewriter/NonAggregationGroupByToDistinctQueryRewriterTest.java:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  @Test
+  // SELECT col1 as col2 FROM foo GROUP BY col1 => SELECT distinct col1 as col2 FROM foo
```
Can we have a test case with multiple expressions and column aliases? Something like: `SELECT DISTINCT col1 as a, col2 as b, concat(col3, col4, '') as c FROM foo`?

xiangfu0 replied:
added.

amrishlal commented on pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/TransformFunctionFactory.java:
```diff
@@ -82,6 +82,7 @@ private TransformFunctionFactory() {
           put(canonicalize(TransformFunctionType.LN.getName().toLowerCase()), LnTransformFunction.class);
           put(canonicalize(TransformFunctionType.SQRT.getName().toLowerCase()), SqrtTransformFunction.class);
 
+          put(canonicalize(TransformFunctionType.AS.getName().toLowerCase()), AsTransformFunction.class);
```
Can we set it up so that user won't be able to invoke AS function directly from SQL, so that they use the AS keyword?

xiangfu0 replied:
no, this is used universally can not be changed

amrishlal commented on pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:
```diff
@@ -161,7 +162,9 @@ private static void validateDistinctQuery(PinotQuery pinotQuery)
         List<Expression> orderByList = pinotQuery.getOrderByList();
         if (orderByList != null) {
-          List<Expression> distinctExpressions = function.getOperands();
+          List<Expression> distinctExpressions = function.getOperands().stream().map(
```
I thought Pinot coding convention didn't allow using "stream()" due to performance concerns?

xiangfu0 replied:
revised the logic

amrishlal commented on pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:
```diff
@@ -580,18 +578,16 @@ private void testDistinctInnerSegmentHelper(String[] queries, boolean isPql) {
+    testDistinctInnerSegmentHelper(new String[]{
+        "SELECT DISTINCT(intColumn, longColumn, floatColumn, doubleColumn, stringColumn, bytesColumn) FROM "
+            + "testTable LIMIT 10000",
+        "SELECT DISTINCT(stringColumn, bytesColumn, floatColumn) FROM testTable WHERE " + "intColumn >= 60 LIMIT 10000",
+        "SELECT DISTINCT(intColumn, rawBytesColumn) FROM testTable ORDER BY " + "rawBytesColumn LIMIT 5",
```
minor: Concatenation of two strings in these two lines can be replaced by a single string.

xiangfu0 replied:
done.

amrishlal commented:
After this change, running the query `select DISTINCT playerID as x, teamID as y from baseballStats limit 10` produces the following results:

![image](https://user-images.githubusercontent.com/4550395/140558893-f19a2a1e-5f03-44d2-957d-8a0c5955baa6.png)

The column names in the result should be `x` instead of `as(playerID,x)` and `y instead of as(teamID,y)` since the columns are being aliased. Can we add a test case for this as well.

xiangfu0 commented:
> After this change, running the query `select DISTINCT playerID as x, teamID as y from baseballStats limit 10` produces the following results:
> 
> ![image](https://user-images.githubusercontent.com/4550395/140558893-f19a2a1e-5f03-44d2-957d-8a0c5955baa6.png)
> 
> The column names in the result should be `x` instead of `as(playerID,x)` and `y instead of as(teamID,y)` since the columns are being aliased. Can we add a test case for this as well.

fixed and added the integration tests

amrishlal commented:
The column names in the resultset look good now, but I think there is more here than meets the eye. During runtime `AsTransformFunction` is being constantly called by `DistinctOperator`. This appears to be redundant since `AsTransformFunction` is simply passing the underlying resultset and has no use during runtime. I think this can be fixed by making the following modification to `BrokerRequestToQueryContextConverter.java`:

```
          case "DISTINCT":
            // Handle alias
            operands = thriftExpression.getFunctionCall().getOperands();
            for (Expression operand : operands) {
              if (operand.isSetFunctionCall() && operand.getFunctionCall().getOperator().equalsIgnoreCase("AS")) {
                List<Expression> asOperands = operand.getFunctionCall().getOperands();
                aliasList.add(asOperands.get(1).getIdentifier().getName());

                // Now that alias has been set, we don't need AS function, so replace AS function with its first
                // operand.
                Expression mainOperand = asOperands.get(0);
                operand.setType(mainOperand.getType());
                operand.unsetFunctionCall();

                switch (mainOperand.getType()) {
                  case LITERAL:
                    operand.setLiteral(mainOperand.getLiteral());
                    break;
                  case FUNCTION:
                    operand.setFunctionCall(mainOperand.getFunctionCall());
                    break;
                  case IDENTIFIER:
                    operand.setIdentifier(mainOperand.getIdentifier());
                    break;
                }
              } else {
                aliasList.add(null);
              }
            }
            break;

```
However, this fix still appears a bit "hacky" since AsTransformFunction is used only during compile time and then discarded. One way to fix this might be to modify `org.apache.pinot.common.request.Function` to contain a list of output aliases (?) Since DISTINCT is a Function that outputs an entire row (and there are probably other functions like this), it makes sense for `Function` to have a list of aliases for the row output as well (?) and this will avoid the use of `AsTransformFunction` and all we would need to do is to fetch the list of aliases from the Function and set it (?)

xiangfu0 commented:
> The column names in the resultset look good now, but I think there is more here than meets the eye. During runtime `AsTransformFunction` is being constantly called by `DistinctOperator`. This appears to be redundant since `AsTransformFunction` is simply passing the underlying resultset and has no use during runtime. I think this can be fixed by making the following modification to `BrokerRequestToQueryContextConverter.java`:
> 
> ```
>           case "DISTINCT":
>             // Handle alias
>             operands = thriftExpression.getFunctionCall().getOperands();
>             for (Expression operand : operands) {
>               if (operand.isSetFunctionCall() && operand.getFunctionCall().getOperator().equalsIgnoreCase("AS")) {
>                 List<Expression> asOperands = operand.getFunctionCall().getOperands();
>                 aliasList.add(asOperands.get(1).getIdentifier().getName());
> 
>                 // Now that alias has been set, we don't need AS function, so replace AS function with its first
>                 // operand.
>                 Expression mainOperand = asOperands.get(0);
>                 operand.setType(mainOperand.getType());
>                 operand.unsetFunctionCall();
> 
>                 switch (mainOperand.getType()) {
>                   case LITERAL:
>                     operand.setLiteral(mainOperand.getLiteral());
>                     break;
>                   case FUNCTION:
>                     operand.setFunctionCall(mainOperand.getFunctionCall());
>                     break;
>                   case IDENTIFIER:
>                     operand.setIdentifier(mainOperand.getIdentifier());
>                     break;
>                 }
>               } else {
>                 aliasList.add(null);
>               }
>             }
>             break;
> ```
> 
> However, this fix still appears a bit "hacky" since AsTransformFunction is used only during compile time and then discarded. One way to fix this might be to modify `org.apache.pinot.common.request.Function` to contain a list of output aliases (?) Since DISTINCT is a Function that outputs an entire row (and there are probably other functions like this), it makes sense for `Function` to have a list of aliases for the row output as well (?) and this will avoid the use of `AsTransformFunction` and all we would need to do is to fetch the list of aliases from the Function and set it (?)

In this case, then we should try to delete the AsTransformFunction and try to compute the alias at the beginning then replace the AS expression to just its left expression.

amrishlal commented:
> > However, this fix still appears a bit "hacky" since AsTransformFunction is used only during compile time and then discarded. One way to fix this might be to modify `org.apache.pinot.common.request.Function` to contain a list of output aliases (?) Since DISTINCT is a Function that outputs an entire row (and there are probably other functions like this), it makes sense for `Function` to have a list of aliases for the row output as well (?) and this will avoid the use of `AsTransformFunction` and all we would need to do is to fetch the list of aliases from the Function and set it (?)
> 
> In this case, then we should try to delete the AsTransformFunction and try to compute the alias at the beginning then replace the AS expression to just its left expression.
> 
> Typical query engine can generate the query plan with the output column names.

ok thanks, LGTM now.
